### DISCUSSION TITLE
How can I instantiate a DataTemplate(VirtualizingWrapPanel)

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#17268] [Q&A] [answered]
Recently I plan to make a VirtualizingWrapPanel. First, I checked the source code of Avalonia WrapPanel, and I could understand the principle of implementation in the source code.

Then I checked the source code of VirtualizingStackPanel to understand the principle of virtualization. But I found that I couldn't understand its source code at all.

Then I tried to observe its principle in Visual Tree, and I could roughly see that its principle is that the IsVisible of the control that is not in the visual range is false, while the IsVisible of the control in the visual range is true.

I continued to look at the source code of VirtualizingStackPanel and VirtualizingPanel in this way, and roughly saw that its principle is to use ItemContainerGenerator to complete the calculation in an ItemsControl first, and then instantiate each item and put it in the Visual Tree when the control is loaded for the first time (but I can't find any instantiate code or code to add items to VirtualizingStackPanel).

But I find it strange that I observed that virtualization is controlled and realized by controlling IsVisible, but I can't find IsVisible related code in VirtualizingStackPanel and VirtualizingPanel. Moreover, there are too many protected properties and methods in VirtualizingStackPanel and VirtualizingPanel, and it is impossible to rewrite them using their source code.

So, I plan to complete this control by using IsVisible to instantiate it according to my own observations. But now I have a problem, when using DataTemplate, for example:
```
<ScrollViewer Name="MyScrollViewer">
	<ItemsControl Name="MyItemsControl" ItemsSource="{Binding TestOC}">
		<ItemsControl.ItemsPanel>
			<ItemsPanelTemplate>
				<Local:VirtualizingWrapPanel></Local:VirtualizingWrapPanel>
			</ItemsPanelTemplate>
		</ItemsControl.ItemsPanel>
		<ItemsControl.ItemTemplate>
			<DataTemplate x:DataType="Local:TestM">
				<Border Height="100" Width="100" Background="{Binding Background}">
					<TextBlock Text="{Binding Content}" Foreground="White" VerticalAlignment="Center" HorizontalAlignment="Center"></TextBlock>
				</Border>
			</DataTemplate>
		</ItemsControl.ItemTemplate>
	</ItemsControl>
</ScrollViewer>
```
When the control is loaded for the first time, I need to instantiate each item according to the DataTemplate and transfer the Model binding to each item. My idea is to keep searching upwards in VirtualizingWrapPanel, find an element such as ItemsControl or ListBox, and then get its DataTemplate. But after getting the DataTemplate, I don't know how to instantiate it. Is there a better way, or is my idea not feasible and should I use other ideas?

**Comment by timunie** [comment, marked as answer]
Check out how VirtualizatibgStackPanel works. Afaik you don't need to initiate the Template on your own

**Reply by kyuranger** [reply]
I tried to imitate the code of VirtualizingStackPanel like this:
```
 private void VirtualizingWrapPanel_Loaded(object? sender, RoutedEventArgs e)
 {
     var generator = ItemContainerGenerator!;
     var container = generator.CreateContainer(Items, 0, null);

     generator.PrepareItemContainer(container, Items, 0);
     AddInternalChild(container);
     generator.ItemContainerPrepared(container, null,0);     
 }
```
But strangely, only a contentpresenter is added to the Visual Tree, and no content is instantiated at all.

**Reply by kyuranger** [reply]
And it only adds one Contentpresenter, not the same number of Contentpresenters as Items.

**Reply by kyuranger** [reply]
I found the problem. I should use a single item to do this instead of all items. Thank you for your advice. I will continue to study it carefully.

**Comment by kyuranger** [comment]
I referred to some of the VirtualizingStackPanel code and implemented VirtualizingWrapPanel in my own way.

This is the GitHub repository for this VirtualizingWrapPanel project:
https://github.com/kyuranger/Avalonia.SVirtualizingWrapPanel/tree/main

And here is the result:
![2](https://github.com/user-attachments/assets/840a6e62-488d-440d-80e6-ed3c901707d8)

**Reply by timunie** [reply]
niiiice ❤️
